packages/daemon/src/service/pipeline.ts

```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
   });
 }
 
+interface RunnerOptions {
+  job: JobEntity;
+  pipeline: PipelineEntity;
+  plugins: Partial<Record<PluginTypeI, PluginInfo>>;
+  tracer: Tracer;
+  runnable: PluginRunnable;
+  datasetRoot: string;
+}
+
+interface ModelResult {
+  model: RunnableResponse;
+  modelPlugin: PluginPackage;
+}
+
+interface RunnerResult {
+  output: any;
+  dataset: any;
+  modelPath: string;
+  modelPlugin: PluginPackage;
+  dataProcess: PluginPackage;
+  datasetProcess: PluginPackage;
+}
+
+/**
+ * job runner
+ */
+export class JobRunner {
+  opts: RunnerOptions;
+  constructor(
+    opts: RunnerOptions) {
+    this.opts = opts;
+  }
+  getParams(params: string | null, ...extra: object[]): object {
+    if (params == null) {
+      return Object.assign({}, ...extra);
+    } else {
+      return Object.assign(JSON.parse(params), ...extra);
+    }
+  }
+  verifyPlugin (name: string): void {
+    if (!this.opts.plugins[name]) {
+      throw new TypeError(`"${name}" plugin is required`);
+    }
+  }
+  dispatchJobEvent(jobStatus: PipelineStatus, step?: PluginTypeI, stepAction?: 'start' | 'end') {
+    const jobEvent = new JobStatusChangeEvent(
+      jobStatus,
+      step,
+      stepAction
+    );
+    this.opts.tracer.dispatch(jobEvent);
+  }
+  async runPlugin(type: PluginTypeI, ...args: any[]) {
+    const plugin = this.opts.plugins[type].plugin;
+    this.dispatchJobEvent(PipelineStatus.RUNNING, type, 'start');
+    const result = await this.opts.runnable.start(plugin, ...args);
+    this.dispatchJobEvent(PipelineStatus.RUNNING, type, 'end');
+    return result;
+  }
+
```

Comment:
Unnecessary empty line.

Reply:
What's the standard style we follow here?

Reply:
You should keep a consistent style for your class.